pulsar-client/src/main/java/org/apache/pulsar/client/impl/ConsumerImpl.java

```diff
@@ -932,8 +935,52 @@ private void increaseAvailablePermits(ClientCnx currentCnx, int delta) {
         }
     }
 
+    private ByteBuf decryptPayloadIfNeeded(MessageIdData messageId, MessageMetadata msgMetadata, ByteBuf payload,
+            ClientCnx currentCnx) {
+
+        if (msgMetadata.getEncryptionKeysCount() == 0) {
+            return payload.retain();
+        }
+
+        // If KeyReader is not configured throw exception based on config param
+        if (conf.getCryptoKeyReader() == null) {
+
+            if (conf.getCryptoFailureAction() == ConsumerCryptoFailureAction.CONSUME) {
+                log.warn("CryptoKeyReader interface is not implemented. Consuming encrypted message.");
+                return payload.retain();
+            } else if (conf.getCryptoFailureAction() == ConsumerCryptoFailureAction.DISCARD) {
+                log.warn(
+                        "Skipping decryption since CryptoKeyReader interface is not implemented and config is set to discard");
+                discardMessage(messageId, currentCnx, ValidationError.DecryptionError);
+            }
+            return null;
+        }
+
+        // Create msgCrypto if not created already
+        if (this.msgCrypto == null) {
+            this.msgCrypto = new MessageCrypto(false);
```

Comment:
though this method will not be called concurrently but as this method is not thread-safe, should we initialize `msgCrypto` in constructor if `conf.getCryptoKeyReader()` present.?